File: packages/server/src/matching-engine.ts
import { v4 as uuidv4 } from 'uuid';
import type { Order, Trade, Result } from '@orderbook/shared';
import type { OrderBook } from './orderbook.js';
import type { PortfolioManager } from './portfolio-manager.js';

/**
 * MatchingEngine
 *
 * Validates incoming orders, runs price-time priority matching,
 * and delegates balance updates to the PortfolioManager.
 */
export class MatchingEngine {
	constructor(
		private orderBook: OrderBook,
		private portfolioManager: PortfolioManager,
	) {}

	/**
	 * Validate & match an incoming order against the book.
	 * Returns executed trades + any remaining (resting) order.
	 */
	placeOrder(
		order: Order,
	): Result<{ trades: Trade[]; remainingOrder: Order | null }> {
		// ── Input validation ───────────────────────────────
		if (order.price <= 0) return { error: 'Price must be positive' };
		if (order.quantity <= 0) return { error: 'Quantity must be positive' };
		if (order.price > 1_000_000) return { error: 'Price too high' };
		if (order.quantity > 10_000) return { error: 'Quantity too high' };

		if (
			!this.portfolioManager.canAfford(
				order.userId,
				order.side,
				order.price,
				order.quantity,
			)
		) {
			return { error: 'Insufficient funds' };
		}

		// ── Matching loop
		const trades: Trade[] = [];
		let remainingQty = order.quantity;

		while (remainingQty > 0) {
			const opposite =
				order.side === 'buy'
					? this.orderBook.getBestAsk()
					: this.orderBook.getBestBid();

			if (!opposite) break;

			const canMatch =
				order.side === 'buy'
					? order.price >= opposite.price
					: order.price <= opposite.price;

			if (!canMatch) break;

			// Execute trade at maker's price
			const tradeQty = Math.min(remainingQty, opposite.quantity);

			const trade: Trade = {
				id: uuidv4(),
				buyOrderId: order.side === 'buy' ? order.id : opposite.id,
				sellOrderId: order.side === 'sell' ? order.id : opposite.id,
				buyerId: order.side === 'buy' ? order.userId : opposite.userId,
				sellerId: order.side === 'sell' ? order.userId : opposite.userId,
				price: opposite.price, // taker gets maker's price
				quantity: tradeQty,
				timestamp: Date.now(),
			};

			trades.push(trade);
			remainingQty -= tradeQty;

			// Update or remove the resting order
			if (opposite.quantity === tradeQty) {
				this.orderBook.removeOrder(opposite.id);
			} else {
				opposite.quantity -= tradeQty;
			}

			this.portfolioManager.executeTrade(trade);
		}

		// Remaining quantity rests on the book
		if (remainingQty > 0) {
			const restingOrder: Order = { ...order, quantity: remainingQty };
			this.orderBook.addOrder(restingOrder);
			return { success: { trades, remainingOrder: restingOrder } };
		}

		return { success: { trades, remainingOrder: null } };
	}

	/** Cancel an existing order. */
	cancelOrder(orderId: string, userId: string): Result<Order> {
		const order = this.orderBook.getOrderById(orderId);
		if (!order) return { error: 'Order not found' };
		if (order.userId !== userId) return { error: 'Not your order' };

		this.orderBook.removeOrder(orderId);
		return { success: order };
	}
}

File: packages/matching-engine-go/internal/matching_engine.go
package internal

import (
	"errors"
	"time"

	"github.com/google/uuid"
)

// MatchingEngine validates and matches orders using price-time priority
type MatchingEngine struct {
	orderBook *OrderBook
}

// NewMatchingEngine creates a new matching engine instance
func NewMatchingEngine() *MatchingEngine {
	return &MatchingEngine{
		orderBook: NewOrderBook(),
	}
}

// PlaceOrder validates and matches an incoming order against the book
// Returns executed trades and any remaining (resting) order
func (me *MatchingEngine) PlaceOrder(order *Order) ([]*Trade, *Order, error) {
	// Input validation
	if order.Price <= 0 {
		return nil, nil, errors.New("price must be positive")
	}
	if order.Quantity <= 0 {
		return nil, nil, errors.New("quantity must be positive")
	}
	if order.Price > 1_000_000 {
		return nil, nil, errors.New("price too high")
	}
	if order.Quantity > 10_000 {
		return nil, nil, errors.New("quantity too high")
	}

	// Note: Portfolio/fund checking is done in Node.js layer
	// This keeps the matching engine focused on pure matching logic

	trades := make([]*Trade, 0)
	remainingQty := order.Quantity

	// Matching loop
	for remainingQty > 0 {
		var opposite *Order

		// Get best opposite order
		if order.Side == SideBuy {
			opposite = me.orderBook.GetBestAsk()
		} else {
			opposite = me.orderBook.GetBestBid()
		}

		// No more matches available
		if opposite == nil {
			break
		}

		// Check if prices cross
		var canMatch bool
		if order.Side == SideBuy {
			canMatch = order.Price >= opposite.Price
		} else {
			canMatch = order.Price <= opposite.Price
		}

		if !canMatch {
			break
		}

		// Execute trade at maker's price (price-time priority)
		tradeQty := min(remainingQty, opposite.Quantity)

		trade := &Trade{
			ID:        uuid.New().String(),
			Price:     opposite.Price, // Taker gets maker's price
			Quantity:  tradeQty,
			Timestamp: time.Now().UnixMilli(),
		}

		// Set buyer/seller based on sides
		if order.Side == SideBuy {
			trade.BuyOrderID = order.ID
			trade.SellOrderID = opposite.ID
			trade.BuyerID = order.UserID
			trade.SellerID = opposite.UserID
		} else {
			trade.BuyOrderID = opposite.ID
			trade.SellOrderID = order.ID
			trade.BuyerID = opposite.UserID
			trade.SellerID = order.UserID
		}

		trades = append(trades, trade)
		remainingQty -= tradeQty

		// Update or remove the resting order
		if opposite.Quantity == tradeQty {
			me.orderBook.RemoveOrder(opposite.ID)
		} else {
			opposite.Quantity -= tradeQty
		}
	}

	// If quantity remains, add to book as resting order
	if remainingQty > 0 {
		restingOrder := &Order{
			ID:        order.ID,
			UserID:    order.UserID,
			Side:      order.Side,
			Price:     order.Price,
			Quantity:  remainingQty,
			Timestamp: order.Timestamp,
		}
		me.orderBook.AddOrder(restingOrder)
		return trades, restingOrder, nil
	}

	return trades, nil, nil
}

// CancelOrder cancels an existing order
func (me *MatchingEngine) CancelOrder(orderID string, userID string) (*Order, error) {
	order := me.orderBook.GetOrderByID(orderID)
	if order == nil {
		return nil, errors.New("order not found")
	}

	if order.UserID != userID {
		return nil, errors.New("not your order")
	}

	removed := me.orderBook.RemoveOrder(orderID)
	return removed, nil
}

// GetOrderBook returns the current order book snapshot
func (me *MatchingEngine) GetOrderBook() *OrderBookSnapshot {
	return &OrderBookSnapshot{
		Bids:      me.orderBook.GetTop10Bids(),
		Asks:      me.orderBook.GetTop10Asks(),
		Timestamp: time.Now().UnixMilli(),
	}
}

// Size returns total active orders in the book.
func (me *MatchingEngine) Size() int {
	return me.orderBook.Size()
}

// Helper function for min
func min(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}
